sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala

```diff
@@ -1432,6 +1695,53 @@ case class Logarithm(left: Expression, right: Expression)
     newLeft: Expression, newRight: Expression): Logarithm = copy(left = newLeft, right = newRight)
 }
 
+trait BaseBinaryExpression extends BinaryExpression
+  with ExpectsInputTypes
+  with Serializable
+  with ImplicitCastInputTypes {
+  val child: Expression
+  val scale: Expression
+  override def left: Expression = child
+  override def right: Expression = scale
+  override def nullable: Boolean = true
+  override def foldable: Boolean = child.foldable
+
+  /**
+   * Expected input types from child expressions. The i-th position in the returned seq indicates
+   * the type requirement for the i-th child.
+   *
+   * The possible values at each position are:
+   *   1. a specific data type, e.g. LongType, StringType.
+   *   2. a non-leaf abstract data type,
+   *      e.g.NumericType, IntegralType, FractionalType.
+   */
+  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, IntegerType)
+
+  // Avoid repeated evaluation since `scale` is a constant int,
+  // avoid unnecessary `child` evaluation in both codegen and non-codegen eval
+  // by checking if scaleV == null as well.
+  protected lazy val scaleV: Any = scale.eval(EmptyRow)
+
+  protected lazy val _scale: Int = scaleV.asInstanceOf[Int]
```

Comment:
the `scaleV` doesn't appear to be referenced anywhere else now. Should we just name this `protected lazy val scaleValue` instead to simplify, dropping the leading underscore?

Reply:
`scaleV` is used in some other udfs `ceil`, `floor`, so just renamed  `_scale` to `scaleValue`